packages/cli/src/decorators/controller.registry.ts

```diff
@@ -0,0 +1,136 @@
+import { userHasScope } from '@/permissions/checkAccess';
+
+import type {
+	AccessScope,
+	Controller,
+	ControllerMetadata,
+	HandlerName,
+	RateLimit,
+	RouteMetadata,
+} from './types';
+
+const registry = new Map<Function, ControllerMetadata>();
+
+export const getControllerMetadata = (controllerClass: Function) => {
+	let metadata = registry.get(controllerClass);
+	if (!metadata) {
+		metadata = {
+			basePath: '/',
+			middlewares: [],
+			routes: new Map(),
+		};
+		registry.set(controllerClass, metadata);
+	}
+	return metadata;
+};
+
+export const getRouteMetadata = (controllerClass: Function, handlerName: HandlerName) => {
+	const metadata = getControllerMetadata(controllerClass);
+	let route = metadata.routes.get(handlerName);
+	if (!route) {
+		route = {} as RouteMetadata;
+		metadata.routes.set(handlerName, route);
+	}
+	return route;
+};
+
+@Service()
+export class ControllerRegistry {
+	constructor(
+		private readonly license: License,
+		private readonly authService: AuthService,
+	) {}
+
+	activate(app: Application) {
+		for (const controllerClass of registry.keys()) {
+			this.activateController(app, controllerClass);
+		}
+	}
+
+	private activateController(app: Application, controllerClass: Function) {
+		const metadata = registry.get(controllerClass)!;
+
+		const router = Router({ mergeParams: true });
+		const prefix = `/${[config.getEnv('endpoints.rest'), metadata.basePath].join('/')}`
+			.replace(/\/+/g, '/')
+			.replace(/\/$/, '');
+		app.use(prefix, router);
+
+		const controller = Container.get<Controller>(controllerClass);
+		const controllerMiddlewares = metadata.middlewares.map(
+			(handlerName) => controller[handlerName].bind(controller) as RequestHandler,
+		);
+
+		for (const [handlerName, route] of metadata.routes) {
+			const handler = async (req: Request, res: Response) =>
+				await controller[handlerName](req, res);
+
+			router[route.method](
+				route.path,
+				...(inProduction && route.rateLimit
+					? [this.createRateLimitMiddleware(route.rateLimit)]
+					: []),
+				// eslint-disable-next-line @typescript-eslint/unbound-method
+				...(route.skipAuth ? [] : [this.authService.authMiddleware]),
+				...(route.licenseFeature ? [this.createLicenseMiddleware(route.licenseFeature)] : []),
+				...(route.accessScope ? [this.createScopedMiddleware(route.accessScope)] : []),
+				...controllerMiddlewares,
+				...route.middlewares,
+				route.usesTemplates ? handler : send(handler),
+			);
+		}
+	}
+
+	private createRateLimitMiddleware(rateLimit: RateLimit): RequestHandler {
+		return expressRateLimit({
+			windowMs: rateLimit.windowMs,
+			limit: rateLimit.limit,
+			message: { message: 'Too many requests' },
+		});
+	}
+
+	private createLicenseMiddleware(feature: BooleanLicenseFeature): RequestHandler {
+		return (_req, res, next) => {
+			if (!this.license.isFeatureEnabled(feature)) {
+				return res
+					.status(403)
+					.json({ status: 'error', message: 'Plan lacks license for this feature' });
```

Comment:
Let's mention the missing feature in the response? Or omitted for security?

Reply:
This is copied for the existing code. I don't have a string opinion on this, but I'm definitely leaning towards not leaking that detail.